src/dune_engine/build_system.ml

```diff
@@ -1253,6 +1253,25 @@ let is_target file =
       in
       loop file)
 
+let list_dir_recursively dir =
+  let parent = Path.Build.parent_exn dir in
+  let* dir_targets = directory_targets_of ~dir:(Path.build parent) in
+  match Path.Set.mem dir_targets (Path.build dir) with
+  | false -> User_error.raise [ Pp.text "not a dir target" ]
+  | true ->
```

Comment:
@snowleopard calling this function to generate the rule is what fails. I suppose I should I just need to build the directory before listing the contents.

Reply:
I'm not sure. I think `directory_targets_of` should always return the correct list, i.e. there should be no need to first build something to get the right result. Otherwise, I guess it's a bug.

Reply:
Hmm, but I don't see how this is possible. The user only specifies that they are producing a directory, not what is going to be in it. So the only way to know what is produced is to build and then inspect.

Reply:
If you have `(targets (dir foo))`, then `foo` is a target directory, but `foo/subdir` is not a target directory. Just like `foo/file` is not a target file. Does this make sense? 

To answer the question "is directory X produced by a rule?" the current implementation asks if "X" is a directory target, then if its parent is a directory target, etc. It's not the fastest possible approach (we can optimise it if needed) but it's simpler compared to requiring to run the build to force `directory_targets_of` to give you the right answer.

Reply:
Oh I think we are talking about a different problem.

The problem is not `directory_targets_of`. The problem is that we want to get all the files (recursively) produced inside a directory target. The only way to get that information is to build the dir first.

Reply:
I'm still confused. Presumably, it's the same when installing file targets: you want to build them first. So, why do directory targets present a problem?

I don't know how things are currently implemented, but in my mind, an installation rule is not very much different from any other rule: it has dependencies (doesn't matter file or directory), which need to be available when it runs, and the only special bit is that the results are put in a special place (rather than into the build directory).

Reply:
This is the distinction between files and directories that I'm trying to draw:

* If we only have files, we only need the list of file paths to compute the rule that creates the .install file. The files themselves are only dependencies for executing the rule, but not for creating the rule.
* If we have directories, then we need to build directories before we can even compute the rule itself. Because the action that writes the .install file now needs to all the contents of these directories.

I suppose the confusing thing here is why I'm even pointing this out. I'm not sure anymore, and I don't think it matters. So feel free to ignore this "point" of mine.

Reply:
> If we have directories, then we need to build directories before we can even compute the rule itself.

I think this is the bit I don't understand. The list of dependencies is known statically (the directory targets, just the paths to their roots), and the action consists of a bunch of statically known `cp -r`. Perhaps, you don't want to use `bash` actions? Then, yes, I agree, we might want a new `copy-dir` action in Dune's action language.

Reply:
Okay, I think I see the source of confusion. This has nothing to do with copying anything to the install dir. I agree this isn't problematic.

Consider this dune file:

```
(install
 (section doc)
 (package foo)
 (directories mydocs))
(rule
 (targets (dir mydocs))
 (action (bash "mkdir mydocs && touch mydocs/{x,y,z}")))
```

We expect the following `foo.install`:

```
doc: [
  "mydocs/x",
  "mydocs/y",
  "mydocs/z"
]
```

Notice how we needed to build `mydocs` to even compute the rule that generates `foo.install`. Currently, doing this isn't necessary. It's sufficient to just walk all the dune files and evaluate their `install` stanzas to generate the rule.

Reply:
Ah, the `.install` file is what I was missing. I have no idea what it is :-) I guess, it must list everything installed. If yes, I agree, directory targets add a fundamentally new source of dynamism here.

Reply:
The .install file is needed to package things for opam. Opam reads the .install file to figure which files belong to a package and should be copied to the opam switch.

Reply:
Actually, wait a sec.

> Notice how we needed to build `mydocs` to even compute the rule that generates `foo.install`.

Is that really true? To merely compute that rule, you don't need to build `mydocs`. The rule could just be "recursively list the contents of all my directory targets and print them out to `.install` file". Isn't that right?

Reply:
Yes, and that is why I thought about introducing a custom action initially. It just seems really odd to add such a super specialized action.

Reply:
I don't think it's odd. If we make directories first class, like files, we need actions that work with directories. Listing a directory is an equivalent of `cat` for working with files.

Reply:
I'm trying to revive this PR to complete the support for directory targets for 3.2, and I do think that a custom action is a fine solution. But the custom extension will need to be represented by a closure `unit -> unit Fiber.t` that is created when generating the install rules.  To do that, I will need to import a limited part of the machinery in https://github.com/ocaml/dune/pull/3809.

Reply:
I'm not sure why you want the custom action extension machinery but I don't have any objections. It just seems fine to me to just add some new actions for working with directories (we can enable them only when the directory targets extension is enabled).

Reply:
Problem is that introducing that action would require moving a bunch of install stuff back to Dune_engine. We don't want to go backwards in our refactoring :)